codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustWriter.kt

```diff
@@ -465,33 +468,63 @@ class RustWriter private constructor(
     }
 
     /**
-     * Generate a wrapping if statement around a field.
-     *
-     * - If the field is optional, it will only be called if the field is present
-     * - If the field is an unboxed primitive, it will only be called if the field is non-zero
-     *
+     * Generate a wrapping if statement around a nullable value.
+     * The provided code block will only be called if the value is not `None`.
      */
-    fun ifSet(shape: Shape, member: Symbol, outerField: String, block: RustWriter.(field: String) -> Unit) {
+    fun ifSome(member: Symbol, value: ValueExpression, block: RustWriter.(value: ValueExpression) -> Unit) {
         when {
             member.isOptional() -> {
-                val derefName = safeName("inner")
-                rustBlock("if let Some($derefName) = $outerField") {
-                    block(derefName)
+                val innerValue = ValueExpression.Reference(safeName("inner"))
+                rustBlock("if let Some(${innerValue.name}) = ${value.asRef()}") {
+                    block(innerValue)
                 }
             }
 
-            shape is NumberShape -> rustBlock("if ${outerField.removePrefix("&")} != 0") {
-                block(outerField)
+            else -> this.block(value)
+        }
+    }
+
+    /**
+     * Generate a wrapping if statement around a primitive field.
+     * The specified block will only be called if the field is not set to its default value - `0` for
+     * numbers, `false` for booleans.
+     */
+    fun ifNotDefault(shape: Shape, variable: ValueExpression, block: RustWriter.(field: ValueExpression) -> Unit) {
+        when (shape) {
+            is FloatShape, is DoubleShape -> rustBlock("if ${variable.asValue()} != 0.0") {
+                block(variable)
+            }
+
+            is NumberShape -> rustBlock("if ${variable.asValue()} != 0") {
+                block(variable)
             }
 
-            shape is BooleanShape -> rustBlock("if ${outerField.removePrefix("&")}") {
-                block(outerField)
+            is BooleanShape -> rustBlock("if ${variable.asValue()}") {
+                block(variable)
             }
 
-            else -> this.block(outerField)
+            else -> rustBlock("") {
+                this.block(variable)
+            }
         }
     }
 
+    /**
+     * Generate a wrapping if statement around a field.
+     *
+     * - If the field is optional, it will only be called if the field is present
+     * - If the field is an unboxed primitive, it will only be called if the field is non-zero
+     *
+     */
+    fun ifSet(
+        shape: Shape,
+        member: Symbol,
+        variable: ValueExpression,
+        block: RustWriter.(field: ValueExpression) -> Unit,
+    ) {
+        ifSome(member, variable) { inner -> ifNotDefault(shape, inner, block) }
+    }
+
```

Comment:
I think we should start including examples of generated code for these so that people can understand how they work at a glance. Just 2 snippets of code for each would be helpful.

Reply:
Does this work? https://github.com/awslabs/smithy-rs/pull/2005/commits/3160a92c17828bceb680b40c7e39c715d512a791